### ISSUE TITLE
epoch이 한 번 밖에 진행되지 않아요

### ISSUE THREAD

**Issue by joonwest** [sooftware/KoSpeech#30]
> [2020-07-03 23:51:30,154 utils.py:21 - info()] Epoch 0 Training result saved as a csv file complete !!

로컬에서 코드 돌려봤는데 이상없이 실행은 되는데 첫번째 epoch이후에 2번째로 넘어가질 않고 위에 코드에서 계속 멈춰있는 상황입니다
왜 이런 상황이 나타나는지 알 수 있을까요? 
그리고 초반에는 train CER val CER의 차이가 있는게 정상인가요?

**Comment by sooftware** [comment]
해당 문제는 데이터셋에 문제있는 오디오 파일 (다운로드가 제대로 되지 않았다거나 하는 등) 이 있는 경우,
스레드 중 하나가 죽어서 생기는 문제로 파악됩니다. (현재까지는 그렇게 파악이 됩니다.)  
  
로그에서 에러가 났었는지를 한번 확인해주세요.  
그리고 초반 train cer 및 valid cer 관련 질문을 해주셨는데, train cer의 경우 누적된 cer의 평균을 나타내도록 해주었기 때문에  
valid cer과 차이가 있습니다. 특히 첫 에폭에는 두드러지게 나게됩니다. (이와 같이 표시한 이유는 cer이 수렴하는지에 대해 보기 쉽도록 하기 위함입니다.)

**Comment by joonwest** [comment]
혹시나 데이터 문제인가 해서 크기가 작은 몇개의 데이터만 가지고 코드를 돌려도 에폭이 넘어가지않네요
말씀해주신 스레드 문제인거 같은데
로그에서 에러가 난것은 어떻게 확인을 할 수있을까요?
터미널에서는 이렇게 나오는 것이 전부에요 

> ../data/checkpoint/checkpoints/2020_07_04_23_17_39/trainer_states.pt
> ../data/checkpoint/checkpoints/2020_07_04_23_17_39/model.pt
> [2020-07-04 23:17:39,539 utils.py:21 - info()] train() completed
> [2020-07-04 23:17:39,650 utils.py:21 - info()] save checkpoints
> ../data/checkpoint/checkpoints/2020_07_04_23_17_39/trainer_states.pt
> ../data/checkpoint/checkpoints/2020_07_04_23_17_39/model.pt
> [2020-07-04 23:17:39,650 utils.py:21 - info()] Epoch 0 (Training) Loss 0.0355 CER 0.9831
> [2020-07-04 23:17:39,650 utils.py:21 - info()] validate() start
> [2020-07-04 23:18:10,891 utils.py:21 - info()] validate() completed
> [2020-07-04 23:18:10,891 utils.py:21 - info()] Epoch 0 (Validate) Loss 1.0000 CER 1.1024
> [2020-07-04 23:18:10,893 utils.py:21 - info()] Epoch 0 Training result saved as a csv file complete !!

**Comment by sooftware** [comment]
한번 코드쪽을 다시 확인해보겠습니다.

**Comment by sooftware** [comment]
혹시 한줄씩 프린트 찍으셔서 어디서 멈추시는지 확인 해주실 수 있을까요?

**Comment by joonwest** [comment]
제가 잘 몰라서 그러는데 한줄씩 프린트 하는게 파이참에서 main.py 파일을 한 줄씩 실행시켜 보는 건가요?
번거롭게해서 죄송합니다 ㅠㅠ

**Comment by sooftware** [comment]
![image](https://user-images.githubusercontent.com/42150335/86534054-6f8b3f00-bf10-11ea-99e9-4cbc0d2b14f0.png)
이런 식으로 수정해서 해보실래요??

**Comment by joonwest** [comment]
감사합니다!

방금 'path' 부분에 checkpoint에 있는 에폭1번만 돌렸던 model.pt를 넣고서 실행했는데
에폭 1번 이후에 넘어가지 않네요...ㅠㅠ

추가로 한가지 질문이 있는데 val loss가 계속 1로 나오는 이유가 있을까요?
에폭한번씩  돌리고 resume 하면서 확인해보면 CER은 떨어져도 val loss는 계속 1로 나오더라고요


> [2020-07-05 23:18:33,325 utils.py:21 - info()] Epoch 0 (Training) Loss 0.0360 CER 1.0000
> [2020-07-05 23:18:33,326 utils.py:21 - info()] validate() start
> [2020-07-05 23:19:16,164 utils.py:21 - info()] validate() completed
> [2020-07-05 23:19:16,165 utils.py:21 - info()] Epoch 0 (Validate) Loss 1.0000 CER 1.0000
> [2020-07-05 23:19:16,169 utils.py:21 - info()] Epoch 0 Training result saved as a csv file complete !!

**Comment by sooftware** [comment]
1에폭 이후 진행되지 않는 문제는 파악되는대로 글 남기겠습니다 ㅜ.ㅜ  
valid loss 같은 경우는 제가 1.0으로 고정해뒀기 때문에 그렇게 나오는 것입니다.  
  
validation에서 CER 이외에는 의미가 없다고 판단하고, 굳이 Loss를 계산하지 않도록 로직을 수정했습니다.  
이전에 구현해뒀던 모듈들과 호환되기 위해 임의로 1.0이 저장되도록 해뒀습니다.

**Comment by ljymill** [comment]
저도 1 epoch 이후로 넘어가지 않아, 디버깅 해본 결과

    supervised_trainer.train_epoches 함수 내에서
       self.num_workers 을 감소하요 0이 되는 문제가 발생하는것으로 확인하였습니다.


아래 처름 코드로 수정하니 정상적으로 동작 하는것을 확인하였습니다.

        num_workers = self.num_workers
        while True:
            inputs, scripts, input_lengths, target_lengths = queue.get()
            if inputs.shape[0] == 0:
                # Empty feats means closing one loader
                num_workers -= 1  # self.num_workers  0으로 감소하여 1 epoch 이후 진행 되지 않음
                logger.debug('left train_loader: %d' % num_workers)

                if num_workers == 0: # self.num_workers => num_workers로 수정
                    break
                else:
                    continue
....

혹시 다른 부분을 수정해야하는지 확인 부탁드립니다.

**Comment by sooftware** [comment]
@ljymill 님 감사합니다.  
요즘 바빠서 코드 디버깅을 못하고 있었는데 찾아주신 덕분에 수정했습니다.  
   
 전에 함수에서 클래스화하다보니 해당 부분을 고려 못했습니다.   
@joonwest 님도 위와 같이 수정 혹은 코드를 새로 clone하셔서 사용하시면 될 것 같습니다
